File: notification-hubs-test-app-java/src/main/java/com/example/notification_hubs_test_app_java/NotificationDetailActivity.java
package com.example.notification_hubs_test_app_java;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class NotificationDetailActivity extends AppCompatActivity {
    private static final String INTENT_KEY_PREFIX = "com.example.notification_hubs_test_app_java.detail:";
    public static final String INTENT_TITLE_KEY = INTENT_KEY_PREFIX + "title";
    public static final String INTENT_BODY_KEY = INTENT_KEY_PREFIX + "body";

    private static final Set<String> RESERVED_KEYS;

    static {
        RESERVED_KEYS = new HashSet<String>();
        RESERVED_KEYS.add(INTENT_TITLE_KEY);
        RESERVED_KEYS.add(INTENT_BODY_KEY);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Intent info = getIntent();

        setContentView(R.layout.activity_notification_detail);

        final TextView titleValue = findViewById(R.id.detailTitleValue);
        titleValue.setText(info.getStringExtra(INTENT_TITLE_KEY));

        final TextView bodyValue = findViewById(R.id.detail_body_view);
        bodyValue.setText(info.getStringExtra(INTENT_BODY_KEY));

        final LinearLayout detailContent = findViewById(R.id.detail_content);
        for (String key : getDataKeys(info)) {
            final View dataRow = getLayoutInflater().inflate(R.layout.data_item, null);
            final TextView dataKey = dataRow.findViewById(R.id.data_key);
            dataKey.setText(key);

            final TextView dataValue = dataRow.findViewById(R.id.data_value);
            dataValue.setText(info.getStringExtra(key));

            detailContent.addView(dataRow);
        }
    }

    private static Iterable<String> getDataKeys(Intent i) {
        List<String> retval = new ArrayList<String>();
        for (String key: i.getExtras().keySet()) {
            if(!RESERVED_KEYS.contains(key)) {
                retval.add(key);
            }
        }

        Collections.sort(retval);
        return retval;
    }
}

File: notification-hubs-sample-app-kotlin/src/main/java/com/example/notification_hubs_test_app_kotlin/NotificationDetailActivity.kt
package com.example.notification_hubs_test_app_kotlin

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class NotificationDetailActivity : AppCompatActivity() {
    companion object {
        private const val INTENT_KEY_PREFIX = "com.example.notification_hubs_test_app_kotlin.detail:"
        const val INTENT_TITLE_KEY = INTENT_KEY_PREFIX + "title"
        const val INTENT_BODY_KEY = INTENT_KEY_PREFIX + "body"

        private var RESERVED_KEYS: HashSet<String> = HashSet()

        private fun getDataKeys(i: Intent): Iterable<String> {
            val retval: ArrayList<String> = ArrayList()
            for (key in i.extras!!.keySet()) {
                if (!RESERVED_KEYS.contains(key)) {
                    retval.add(key)
                }
            }
            retval.sort()
            return retval
        }

        init {
            RESERVED_KEYS.add(INTENT_TITLE_KEY)
            RESERVED_KEYS.add(INTENT_BODY_KEY)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val info: Intent = intent
        setContentView(R.layout.activity_notification_detail)
        val titleValue: TextView = findViewById(R.id.detailTitleValue)
        titleValue.text = info.getStringExtra(INTENT_TITLE_KEY)
        val bodyValue: TextView = findViewById(R.id.detail_body_view)
        bodyValue.text = info.getStringExtra(INTENT_BODY_KEY)
        val detailContent: LinearLayout = findViewById(R.id.detail_content)
        for (key in getDataKeys(info)) {
            val dataRow: View = layoutInflater.inflate(R.layout.data_item, null)
            val dataKey: TextView = dataRow.findViewById(R.id.data_key)
            dataKey.text = key
            val dataValue: TextView = dataRow.findViewById(R.id.data_value)
            dataValue.text = info.getStringExtra(key)
            detailContent.addView(dataRow)
        }
    }
}
